simulation/chart.py

```diff
@@ -1,232 +1,427 @@
-        x_range (Tuple[float,float], optional): Constant range(min,max value) of x axis. Defaults to None.
-        x_log (bool, optional): Use logarithmic scale to display x axis. Defaults to None.
-        y_label (str, optional): Label text to display next to y axis. Defaults to None.
-        y_range (Tuple[float,float], optional): Constant range(min,max value) of y axis. Defaults to None.
-        y_log (bool, optional): Use logarithmic scale for y axis. Defaults to None.
-    Returns:
-        PlotItem: a chart to contain plots
+class IndexedColor(NamedTuple):
+    """Color from a single index. Useful for stepping through a predefined list of colors.
+
+    Attributes:
+        index (int): predefined color index
+        hues (int): number of steps to generate
     """
-    chart = layout.addPlot(title=title)
-    if height is not None:
-        chart.setPreferredHeight(300)
-    legend = chart.addLegend(offset=None)
-    vb: pg.ViewBox = layout.addViewBox()
-    if legend_width is not None:
-        vb.setFixedWidth(legend_width)
-    legend.setParentItem(vb)
-    legend.anchor((0, 0), (0, 0))
-    chart.setClipToView(True)
 
-    if x_label is not None:
-        chart.setLabel("bottom", x_label)
-    if x_log is not None:
-        chart.setLogMode(x=x_log)
-    if x_range is not None:
-        chart.setXRange(*x_range)
+    index: int
+    hues: int
 
-    if y_label is not None:
-        chart.setLabel("left", y_label)
-    if y_log is not None:
-        chart.setLogMode(y=y_log)
-    if y_range is not None:
-        chart.setYRange(*y_range)
 
-    layout.nextRow()
+Color = Union[RGBAColor, NamedColor, IndexedColor, int, float, str]
+"""Color based on model, predefined values, single greyscale value (0.0 - 1.0) , color index or string representing color (“#RGB”,“#RGBA”,“#RRGGBB”,“#RRGGBBAA”)."""
 
-    return chart
 
+@singledispatch
+def _make_color(color: RGBAColor):  # pyright:ignore [reportGeneralTypeIssues]
+    """Convert RGBA color value to QColor"""
+    if color.A is None:
+        return pg.mkColor((color.R, color.G, color.B))
+    else:
+        return pg.mkColor((color.R, color.G, color.B, color.A))
 
-def add_line_plot(
-    chart: pg.PlotItem, name: str, color=None, width: Optional[float] = None, style=None
-) -> pg.PlotDataItem:
-    """Adds a line plot to the chart.
 
-    Args:
-        name (str): Name (legend title) of the plot
-        color (Any, optional): Line color of the plot. Defaults to None.
-        width (float, optional): Width of the plot line. Defaults to None.
-        style (Any, optional): Style of the plot line. Defaults to None.
-    """
-    config = {}
-    if color is not None:
-        config["color"] = color
+@singledispatch
+def _make_color(color: IndexedColor):  # pyright:ignore [reportGeneralTypeIssues]
+    """Convert Indexed color value to QColor"""
+    return pg.mkColor((color.index, color.hues))
+
+
+@singledispatch
+def _make_color(color: Union[int, float, str, None]):
```

Comment:
Seems like two different cases here. One case is if `color = None`, in which case return `None`. The other is if `color = Union[int, float, str]`, in which case, return `pg.mkColor(color)`. Seems like these should be split out into different methods.

Reply:
This `color = None` check is unnecessary as we have already defined default color in all cases places. On the other hand I've tried to build the single dispatch method with `NoneType,() and None` which is not working. Can you suggest a method signature for this ?

Reply:
The documentation has an example using `None`. https://peps.python.org/pep-0443/

Reply:
pyright is not happy with this "Function declaration _make_color is obscured by a declaration of the same name"

``` python
@fun.register(type(None))
def _make_color(arg):
    return None```